File: src/main/java/io/github/hiiragi283/HTMaterialsAPIImpl.java
package io.github.hiiragi283;

import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;

import io.github.hiiragi283.api.HTMaterialsAPI;
import io.github.hiiragi283.api.material.HTMaterialRegistry;
import io.github.hiiragi283.api.part.HTPartManager;
import io.github.hiiragi283.api.shape.HTShapeRegistry;

public final class HTMaterialsAPIImpl implements HTMaterialsAPI {

    static HTShapeRegistry shapeRegistry;
    static HTMaterialRegistry materialRegistry;
    static CreativeTabs creativeTabs;
    static Item iconItem;
    static Item dictionaryItem;
    static HTPartManager partManager;

    @Override
    public HTShapeRegistry shapeRegistry() {
        return shapeRegistry;
    }

    @Override
    public HTMaterialRegistry materialRegistry() {
        return materialRegistry;
    }

    @Override
    public CreativeTabs creativeTab() {
        return creativeTabs;
    }

    @Override
    public Item iconItem() {
        return iconItem;
    }

    @Override
    public Item dictionaryItem() {
        return dictionaryItem;
    }

    @Override
    public HTPartManager partManager() {
        return partManager;
    }
}

File: src/main/kotlin/io/github/hiiragi283/common/core/HTMaterialsAPIImpl.kt
package io.github.hiiragi283.common.core

import io.github.hiiragi283.api.HTLogger
import io.github.hiiragi283.api.HTMaterialsAPI
import io.github.hiiragi283.api.HTMaterialsPlugin
import io.github.hiiragi283.api.item.HTMaterialItemProvider
import io.github.hiiragi283.api.material.HTMaterialRegistry
import io.github.hiiragi283.api.part.HTPartManager
import io.github.hiiragi283.api.shape.HTShapeKey
import io.github.hiiragi283.api.shape.HTShapeRegistry
import net.minecraftforge.common.MinecraftForge

internal object HTMaterialsAPIImpl : HTMaterialsAPI {

    override var materialRegistry: HTMaterialRegistry = HTMaterialRegistry.EMPTY
        internal set
    override var shapeRegistry: HTShapeRegistry = HTShapeRegistry.EMPTY
        internal set
    override var partManager: HTPartManager = HTPartManager.EMPTY
        internal set
    override var materialItemMap: Map<HTShapeKey, HTMaterialItemProvider> = emptyMap()
        internal set

    private val addons: Iterable<HTMaterialsPlugin> by lazy {
        buildList {
            MinecraftForge.EVENT_BUS.post(HTMaterialsPlugin.RegisterEvent(this))
            HTLogger.log { it.info("RegisterEvent fired!") }
        }
    }

    override fun forEachAddon(action: (HTMaterialsPlugin) -> Unit) {
        addons.forEach(action)
    }
}